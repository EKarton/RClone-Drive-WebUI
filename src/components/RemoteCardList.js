import RemoteCard from './RemoteCard';
import './RemoteCardList.scss';

export default function RemoteCardList({ remotes, onClick }) {
  const handleClick = (remote) => () => {
    onClick(remote);
  };

  return (
    <div className="remote-card-list">
      {remotes.sort().map((remote) => (
        <RemoteCard
          key={remote}
          className="remote-card-list__item"
          remote={remote}
          onClick={handleClick(remote)}
        />
      ))}
    </div>
  );
}
