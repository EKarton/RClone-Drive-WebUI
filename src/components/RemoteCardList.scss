.remote-card-list {
  display: flex;
  flex-wrap: wrap;
  gap: 1rem;

  &__item {
    flex-grow: 1;
  }
}
