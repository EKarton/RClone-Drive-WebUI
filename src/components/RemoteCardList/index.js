import RemoteCard from 'components/RemoteCard';
import './index.scss';

export default function RemoteCardList({ remotes, onClick }) {
  const handleClick = (remote) => () => {
    onClick(remote);
  };

  return (
    <div className="remote-card-list">
      {remotes.sort().map((remote) => (
        <RemoteCard
          key={remote}
          className="remote-card-list__item"
          remote={remote}
          onClick={handleClick(remote)}
        />
      ))}
    </div>
  );
}
