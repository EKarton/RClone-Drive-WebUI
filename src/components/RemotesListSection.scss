.remotes-list-section {
  &__header-text {
    padding-bottom: 1rem;
    font-weight: bold;
  }
}
