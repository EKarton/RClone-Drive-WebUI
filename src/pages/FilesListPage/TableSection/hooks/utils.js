import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
