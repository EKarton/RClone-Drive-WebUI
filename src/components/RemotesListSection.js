import Typography from '@mui/material/Typography';
import RemoteCardList from 'components/RemoteCardList';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';
import { StatusTypes } from 'utils/constants';
import './RemotesListSection.scss';

/**
 * Represents the list of remotes with header
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    throw error;
  }

  const handleRemoteCardClicked = (remote) => {
    onRemoteCardClicked(remote);
  };

  return (
    <>
      <Typography
        variant="h7"
        component="div"
        color="text.primary"
        className="remotes-list-section__header-text"
      >
        Your Remotes
      </Typography>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
