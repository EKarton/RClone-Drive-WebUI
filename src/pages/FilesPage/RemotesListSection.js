import { useHistory } from 'react-router';
import { hashRemotePath } from 'utils/remote-paths-url';
import RemoteCardList from 'components/RemoteCardList';
import { StatusTypes } from 'utils/constants';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';

export default function RemotesListSection() {
  const history = useHistory();
  const { status, data: remotes, error } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    return <div>{error.message}</div>;
  }

  const handleRemoteCardClicked = (remote) => {
    history.push(`/files/${hashRemotePath(`${remote}:`)}`);
  };

  return (
    <>
      <h4>Your Remotes</h4>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
