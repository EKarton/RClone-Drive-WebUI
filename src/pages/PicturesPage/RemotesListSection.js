import RemoteCardList from 'components/RemoteCardList';
import { StatusTypes } from 'utils/constants';
import useFetchRemotes from 'hooks/fetch-data/useFetchRemotes';

/**
 * Represents the list of remotes in the Pictures page
 */
export default function RemotesListSection({ onRemoteCardClicked }) {
  const { status, data: remotes } = useFetchRemotes();

  if (status === StatusTypes.LOADING) {
    return null;
  }

  if (status === StatusTypes.ERROR) {
    return 'Error!';
  }

  const handleRemoteCardClicked = (remote) => {
    onRemoteCardClicked(remote);
  };

  return (
    <>
      <h4>Your Remotes</h4>
      <RemoteCardList remotes={remotes} onClick={handleRemoteCardClicked} />
    </>
  );
}
