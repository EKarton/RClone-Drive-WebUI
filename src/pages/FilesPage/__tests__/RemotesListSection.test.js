import RemoteCardList from 'components/RemoteCardList';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { hashRemotePath } from 'utils/remote-paths-url';
import { customRender, waitFor } from 'test-utils/react';
import RemotesListSection from '../RemotesListSection';

jest.mock('components/RemoteCardList');
jest.mock('hooks/rclone/useRCloneClient');

describe('RemotesListSection', () => {
  const fetchRemotesFn = jest.fn();

  beforeEach(() => {
    useRCloneClient.mockReturnValue({
      fetchRemotes: fetchRemotesFn,
    });
  });

  it('should match snapshot given fetching list of remotes were successful', async () => {
    RemoteCardList.mockImplementation(({ remotes }) => {
      return <div>{remotes.join(',')}</div>;
    });
    fetchRemotesFn.mockResolvedValue(['googledrive', 'onedrive', 'icloud']);

    const component = customRender(<RemotesListSection />);

    await waitFor(() => {
      expect(component.baseElement).toMatchSnapshot();
    });
  });

  it('should go to the files page when user clicks on a remote', async () => {
    // Mock the timer to prevent calling onClick immediately
    jest.useFakeTimers();

    RemoteCardList.mockImplementation(({ remotes, onClick }) => {
      setTimeout(() => onClick('googledrive'), 10000);
      return <div>{remotes.join(',')}</div>;
    });

    fetchRemotesFn.mockResolvedValue(['googledrive', 'onedrive', 'icloud']);

    const component = customRender(<RemotesListSection />);

    // Mimic user clicking on the buttotn
    jest.runAllTimers();

    const expectedPath = `/files/${hashRemotePath('googledrive:')}`;
    await waitFor(() => {
      expect(component.history.location.pathname).toEqual(expectedPath);
    });
  });

  it('should show an error page when fetching remotes failed', async () => {
    fetchRemotesFn.mockRejectedValue(new Error('Error!'));

    const component = customRender(<RemotesListSection />);

    await waitFor(() => {
      expect(component.getByText('Error!')).toBeInTheDocument();
      expect(component.baseElement).toMatchSnapshot();
    });
  });
});
