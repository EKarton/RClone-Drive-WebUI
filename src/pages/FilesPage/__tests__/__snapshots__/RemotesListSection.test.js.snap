// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemotesListSection should match snapshot given fetching list of remotes were successful 1`] = `
<body>
  <div />
</body>
`;

exports[`RemotesListSection should show an error page when fetching remotes failed 1`] = `
<body>
  <div>
    <div>
      Error!
    </div>
  </div>
</body>
`;
